reco_utils/recommender/geoimc/GeoIMCdata.py

```diff
@@ -0,0 +1,162 @@
+        self.data_indices = None
+        self.entity_indices = [None, None]
+
+
+    def get_data(self):
+        if self.data_indices is None:
+            return self.data
+        return self.data[self.data_indices]
+
+
+    def get_entity(self, of="row"):
+        idx = 0 if of=="row" else 1
+        if self.entity_indices[idx] is None:
+            return self.entities[idx]
+        return self.entities[idx][self.entity_indices[idx]]
+
+
+class Dataset():
+    """
+    Base class that holds necessary (minimal) information needed
+    """
+
+    def __init__(
+            self,
+            name,
+            features_dim=0,
+            normalize=False,
+            target_transform=''
+    ):
+        """Initialize parameters
+
+        Args:
+            name (str): Name of the dataset
+            features_dim (uint): Dimension of the features. If not 0, PCA is performed
+                on the features as the dimensionality reduction technique
+            normalize (bool): Normalize the features
+            target_transform (str): Transform the target values. Current options are
+                'normalize' (Normalize the values), '' (Do nothing), 'binarize' (convert
+                the values using a threshold defined per dataset)
+
+        """
+        self.name = None
+        self.training_data = None
+        self.test_data = None
+        self.entities = None
+
+        self.features_dim = features_dim
+        self.feat_normalize = normalize
+        self.target_transform = target_transform
+
+
+    def load_data(self, path):
+        raise NotImplementedError(f"{self.name} should implement it")
+
+
+    def binarize_rating(self, v):
+        return 1 if v >=3 else 0
+
+
+    def normalize(self):
+        if self.feat_normalize:
+            for i in range(len(self.entities)):
+                if isspmatrix_csr(self.entities[i]):
+                    print(f"Normalizing CSR matrix")
+                    self.entities[i] = normalize(self.entities[i])
+                else:
+                    self.entities[i] = length_normalize(self.entities[i])
+
+
+    def generate_train_test_data(self, data, test_ratio=0.3):
+        self.training_data = DataPtr(data, self.entities)
+        self.test_data = DataPtr(data, self.entities)
+
+        self.training_data.data_indices, self.test_data.data_indices = train_test_split(
+            np.array(range(0, data.shape[0])),
+            test_size=test_ratio,
+            shuffle=True,
+            random_state=0
+        )
+        self.training_data.entity_indices[0] = self.training_data.data_indices
+        self.test_data.entity_indices[0] = self.test_data.data_indices
+
+
+    def reduce_dims(self):
+        if self.features_dim != 0:
+            self.entities[0] = reduce_dims(self.entities[0], self.features_dim)
+            self.entities[1] = reduce_dims(self.entities[1], self.features_dim)
+            print(f"Dimensionality reduced ...")
+
+
+class ML_100K(Dataset):
```

Comment:
there is only a class for ml100k, would it be possible to add a generalistic class for any arbitrary dataset or at least for all the movielens datasets?

Reply:
I think it should be possible for ML1M in the class of MovieLens datasets, since, ML10M, ML20M don't come with side information (features of the entities). 

Regarding the generic class, I was hoping that `Dataset` and `DataPtr` classes should be sufficient since, all a dataset handler should do is, just load the `DataPtr training_data`, `DataPtr test_data` variables by inheriting the `Dataset` class as per their dataset's train/test split.